### PR TITLE
WasmThemis: Secure Session support

### PR DESCRIPTION
**Pull request by ilammy**
Initial implementation of **Secure Session** wrapper. It is exported as `SecureSession` object.

This is the most complex API of Themis both in implementation and usage. It stems from the fact that it use C callbacks, which do not map onto other languages directly. Our approach for Secure Session is described in code comments as it relies on a number of assumptions. Ultimately, we use Emscripten's `addFunction` API to bridge JavaScript callbacks provided by WasmThemis users into Emscripten world.

Just like Secure Comparator, Secure Session also requires manual resource management from the users. That's unfortunate but unavoidable. Similarly, we provide some additional error checks and reporting to prevent common mistakes in Secure Session usage.

Note that we provide only buffer-oriented API of Secure Session. The only callback that the user can register is the required one for obtaining public key by session ID. Using transport-oriented API in JavaScript it hard because it assumes synchronous IO while normally JavaScript uses asynchronous network transfers.

### CODE REVIEW

File: src/wrappers/themis/wasm/src/secure_session.js [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+ * @file
+ * Themis Secure Session.
+ */
+
+const libthemis = require('./libthemis.js')
+const keygen = require('./secure_keygen.js')
+const errors = require('./themis_error.js')
+const utils = require('./utils.js')
+
+const subsystem = 'SecureSession'
```
Comment (Lagovas) [inline src/wrappers/themis/wasm/src/secure_session.js]: why subsystem? it's commonly used name?) looks like some module, not like name. can we rename to `subsystem_name` or `module_name`?

File: src/wrappers/themis/wasm/src/secure_session.js [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+//          get_public_key_for_id_callback  get_public_key_for_id;
+//
+//          void *user_data;
+//     };
+//
+// On Emscripten target all pointers take up 4 bytes and are aligned at 4-byte boundary.
+// Hence the following sizes and offsets:
+
+const sizeof_secure_session_user_callbacks_t = 20
+const offsetof_get_public_key_for_id         = 12
```
Comment (Lagovas) [inline src/wrappers/themis/wasm/src/secure_session.js]: can you order by offset? 20 -> 16 ->12? it's simplify reading

File: src/wrappers/themis/wasm/src/secure_session.js (hunk 1 above)
Comment (ilammy) [reply src/wrappers/themis/wasm/src/secure_session.js]: Mmm... IIRC, I used this name because the documentation refers to Secure Cell, Secure Message, etc. collectively as [_cryptosystems_](https://docs.cossacklabs.com/pages/cryptosystems/#themis-cryptosystems). No idea why 'sub-' though. That's also the parameter name used in themis_error.js. 

Would `cryptosystem_name` be okay? I'll rename it here and in all other PRs as well.

Comment (Lagovas) [reply src/wrappers/themis/wasm/src/secure_session.js]: it's okay. `subsystem` looks like system library to work with some low-level api of some module, not like just name of currect module )

File: src/wrappers/themis/wasm/src/secure_session.js [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+        let session_id_ptr, private_key_ptr, callbacks_ptr
+        try {
+            session_id_ptr = libthemis._malloc(sessionID.length)
+            private_key_ptr = libthemis._malloc(privateKey.length)
+            callbacks_ptr = libthemis._malloc(sizeof_secure_session_user_callbacks_t)
+            if (!session_id_ptr || !private_key_ptr || !callbacks_ptr) {
+                throw new ThemisError(cryptosystem_name, ThemisErrorCode.NO_MEMORY)
+            }
+            libthemis.writeArrayToMemory(sessionID, session_id_ptr)
+            libthemis.writeArrayToMemory(privateKey, private_key_ptr)
```
Comment (vixentael) [inline src/wrappers/themis/wasm/src/secure_session.js]: clean up on error?

File: src/wrappers/themis/wasm/src/secure_session.js [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+// global registry in JavaScript is the only feasible choice because JavaScript object
+// memory is managed completely opaquely.
+
+var SecureSessionCallbackRegistry = {}
+
+function registerSecureSession(callbacksPtr, session) {
+    SecureSessionCallbackRegistry[callbacksPtr] = session
+}
+
+function unregisterSecureSssion(callbacksPtr) {
```
Comment (vixentael) [inline src/wrappers/themis/wasm/src/secure_session.js]: fix typo

```suggestion
function unregisterSecureSession(callbacksPtr) {
```

File: src/wrappers/themis/wasm/src/secure_session.js [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+
+    destroy() {
+        let status = libthemis._secure_session_destroy(this.sessionPtr)
+        if (status != ThemisErrorCode.SUCCESS) {
+            throw new ThemisError(cryptosystem_name, status,
+                'failed to destroy Secure Session')
+        }
+        this.sessionPtr = null
+
+        unregisterSecureSssion(this.callbacksPtr)
```
Comment (vixentael) [inline src/wrappers/themis/wasm/src/secure_session.js]: fix typo

```suggestion
        unregisterSecureSession(this.callbacksPtr)
```

File: src/wrappers/themis/wasm/test/test.js [hunk 6]
```diff
@@ -77,4 +77,161 @@ describe('wasm-themis', function() {
...
+            assert.throws(() => server.wrap(emptyArray),
+                expectError(ThemisErrorCode.INVALID_PARAMETER)
+            )
+
+            client.destroy()
+            server.destroy()
+        })
+        it('does not allow public keys in constructor', function() {
+            let keyPair = new themis.KeyPair()
+            assert.throws(() => new themis.SecureSession(clientID, keyPair.publicKey, function(){}),
```
Comment (vixentael) [inline src/wrappers/themis/wasm/test/test.js]: add test with empty `clientID` or empty key

File: src/wrappers/themis/wasm/src/secure_session.js (hunk 4 above)
Comment (ilammy) [reply src/wrappers/themis/wasm/src/secure_session.js]: All hail autocomplete!

**Comment by ilammy** [comment]
Memory zeroization and more unit tests for type checks are coming in separate PRs (linked above).

File: src/wrappers/themis/wasm/test/test.js (hunk 6 above)
Comment (ilammy) [reply src/wrappers/themis/wasm/test/test.js]: There's a test for an empty client ID:

https://github.com/cossacklabs/themis/blob/5c01c86dd5b674a9b790c5ebd182f97904b53675/src/wrappers/themis/wasm/test/test.js#L169-L174

The users are also not allowed to pass empty keys. We expect only `PrivateKey` instances and they cannot be constructed from an empty array:

https://github.com/cossacklabs/themis/blob/5c01c86dd5b674a9b790c5ebd182f97904b53675/src/wrappers/themis/wasm/test/test.js#L59-L62

Comment (vixentael) [reply src/wrappers/themis/wasm/test/test.js]: can we test that private keys cannot be constructed from empty strings?

Comment (ilammy) [reply src/wrappers/themis/wasm/test/test.js]: > can we test that private keys cannot be constructed from empty strings?

We do:

https://github.com/ilammy/themis/blob/5c01c86dd5b674a9b790c5ebd182f97904b53675/src/wrappers/themis/wasm/test/test.js#L69-L72

They cannot be constructed from _any_ string, not just empty ones.

I can add a test for empty string specifically, if you wish.
